### PR TITLE
feat!: Implement Requirement 1.1.3

### PR DESCRIPTION
**Pull request by maxveldink**
## This PR

Only the last four commits will be merged in with this PR. The others are waiting on #77 and #78.

- *Breaking*: Renamed `API.provider=(provider)` to `API.set_provider(provider, domain: nil)` to better adhere to the specification. I decided to move away from the `attr_writer` naming convention as the `domain` parameter would be surprising. I would love others' thoughts on this.
- Provider instances are now stored inside a provider hash in the config instance, with the default provider using `nil` as the key. The other option I tried here was to store `@default_provider` in addition to the hash, but this approach was simpler to the slight detriment of readability.
- Refactored `API.provider` to receive `domain` as a keyword argument. If no domain is provided, it will return the default provider. This is backward compatible.
- Implemented Requirement 1.1.3 from spec version 0.7.0.

### Follow-up Tasks
In subsequent PRs, I'll continue on the flag evaluation specifications.

### How to test
Open a console and play around with the `set_provider` method on the API. If no name is provided, the default provider will be replaced. If a name is provided, the provider will be registered using that name. If the same name is provided, the provider will be overridden.

### CODE REVIEW

File: lib/open_feature/sdk/configuration.rb [hunk 1]
```diff
@@ -14,24 +14,28 @@ class Configuration
...
       #   3. Finally, set the internal provider to the new provider
-      def provider=(provider)
-        @provider.shutdown if @provider.respond_to?(:shutdown)
+      def set_provider(provider, domain: nil)
+        @providers[domain].shutdown if @providers[domain].respond_to?(:shutdown)
 
         provider.init if provider.respond_to?(:init)
 
-        @provider = provider
+        @providers[domain] = provider
```
Comment (josecolella) [inline lib/open_feature/sdk/configuration.rb:38]: question: If the domain is nil, we'd be setting the key as nil here, is this what we want?

Comment (maxveldink) [reply lib/open_feature/sdk/configuration.rb:38]: Yep, that was intentional! The default domain would be represented by `nil`.

**Review by beeme1mr — APPROVED** [review]

**Comment by maxveldink** [comment]
@open-feature/sdk-ruby-maintainers Would love to get this merged in soon as well if anyone has some time for a review!

**Comment by beeme1mr** [comment]
I updated the PR title to mark it as a breaking change.

**Pull request merged**